doc/design/service-discovery.md

```diff
@@ -0,0 +1,228 @@
+
+### Status quo in iceoryx Almond
+
+#### Runtime
+
+* `cxx::expected<InstanceContainer, FindServiceError> PoshRuntime::findService(capro::ServiceDescription)`
+  * Sends message over IPC channel
+
+#### RouDi
+
+* `runtime::IpcMessage {ProcessManager,PortManager}::findService(capro::ServiceDescription)`
+  * Called after RouDi has received a request from the runtime
+  * Sends back answer to `PoshRuntime` over IPC channel
+
+## Terminology
+
+## Requirements
+
+* Polling and event notification approach shall both be possible
+
+### Considerations and use-cases
+
+#### AUTOSAR Adaptive
+
+* findService() call as implemented in Almond v1.0
+* ara::com service discovery callbacks
+
+1. `StartFindService()`: Register callback, std::function interface
+1. `StopFindService()`: Unregister callback
+1. `FindService()`: one-shot, polling, synchronous
+
+#### SOME/IP-SD
+
+* Very similar to AUTOSAR Adaptive
+
+#### Cyclone DDS and ROS 2
+
+* Cyclone DDS calls this feature [topic discovery](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/topic_discovery.md). See also the [feature request](https://github.com/eclipse-cyclonedds/cyclonedds/issues/42).
+
+* Based on SDEP (simple endpoint discovery protocol)
+* `dds_ktopic`
+  * stores pointer to `ddsi_topic_definition`
+* Has built-in DCPSTopic topic and contains
+  * Topic name
+  * Type name
+  * Topic key
+  * QoS settings
+* `dds_find_topic_locally` function is available for users
+
+#### Static discovery
+
+* Static discovery shall be supported
+  * Having RouDi read in a config file, which has
+    * All the topics listed
+    * Wiring pub/sub information
+  * During init phase pub/sub are connected
+* Create abstract interface for `ServiceRegistry`
+  * Add new class to `StaticServiceRegistry::StaticServiceRegistry(std::map<CaproIdString_t, instance_t> map)`
+    * No notification-based callback possibilty
+    * Just static, polling with `FindService()` possible
+
+#### Sychronous, one-shot RPC (Polling)
+
+##### Alternative A: Request/ response feature usage
+
+* Re-use [request/ response feature](https://github.com/eclipse-iceoryx/iceoryx/issues/27)
+
+Pro:
+
+* Re-use building blocks to avoid code duplication
+* Dogfooding
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared for a simple RPC call
+
+##### Alternative B: IPC channel usage
+
+Stick with IPC channel
+
+Pro:
+
+* No code changes needed
+
+Con:
+
+* No userspace solution
+  * IPC channel uses OS functions
+* Upper frame limit may lead to segregation into several frames
+
+#### Event-based notification
+
+##### Alternative A: Shared memory + IPC channel
+
+* Send ID's + ABA-counter/ServiceDiscoveryChangeCounter over IPC channel
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it
+
+Pro:
+
+* Fast data transfer
+
+Contra:
+
+* More complex than todays solution
```

Comment:
I disagree, in the end it would be much simpler than todays solution. Since this requires much less code, loops etc. It would also simplify and reduce the user code since the interfaceports would be gone.

Reply:
I disagree, too. Writing a custom thread-safe and lock-free building block bespoke just for this use-case is complex task. In the other alternatives we would use well established patterns req/resp + Listner or `InterfacePort`s.